modules/renter/repair.go

```diff
@@ -154,19 +237,38 @@ func (r *Renter) threadedRepairUploads() {
 			}
 
 			// build host list
+			totalsize := f.pieceSize * uint64(f.erasureCode.NumPieces()) * f.numChunks()
 			var hosts []uploader
-			randHosts := r.hostDB.RandomHosts(f.erasureCode.NumPieces())
-			for i := range randHosts {
+			randHosts := r.hostDB.RandomHosts(f.erasureCode.NumPieces() * 2)
+			for _, h := range randHosts {
+				// probabilistically filter out known bad hosts
+				// unresponsive hosts will be selected with probability 1/(1+nFailures)
+				nFailures, ok := r.blacklist[h.IPAddress]
+				if n, _ := crypto.RandIntn(1 + nFailures); ok && n != 0 {
+					continue
+				}
+
 				// TODO: use smarter duration
-				hostUploader, err := r.newHostUploader(randHosts[i], f.size, defaultDuration, f.masterKey)
+				hostUploader, err := r.newHostUploader(h, totalsize, defaultDuration, f.masterKey)
 				if err != nil {
+					// penalize unresponsive hosts
+					if strings.Contains(err.Error(), "timeout") {
```

Comment:
would much prefer to whitelist errors than blacklist them. We really should be adding hosts to a blacklist if they cause errors.

Only time we shouldn't add them to a blacklist in the case of an error:

1. It's our fault, not theirs
2. There's no way for a dishonest host to trigger that error - it can only be triggered internally

#2 is really important for security.

We want to ignore errors for the two following cases:

1. txn daisy chain is too long
2. wallet, due to restarting, is attempting double spends

I don't think there are other scenarios that should be whitelisted. Those two can be prevented by having the wallet:

1. return an error if a long txn chain is detected
2. have some persistence for txns it makes

I'm guessing we can't get this by 0.4.4, but I hope you agree with the long term need for whitelisting errors as opposed to blacklisting errors.